File: interop/Interop.java
import peergos.server.Builder;
import peergos.shared.Crypto;
import peergos.shared.NetworkAccess;
import peergos.shared.crypto.hash.Blake3;
import peergos.shared.login.mfa.MultiFactorAuthMethod;
import peergos.shared.login.mfa.MultiFactorAuthResponse;
import peergos.shared.social.FollowRequestWithCipherText;
import peergos.shared.crypto.symmetric.SymmetricKey;
import peergos.shared.user.*;
import peergos.shared.user.fs.*;
import peergos.shared.user.fs.archive.*;
import peergos.shared.util.*;

import java.net.URL;
import java.nio.file.*;
import java.security.MessageDigest;
import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.CompletableFuture;

/** Java half of the interop run: checks what the Rust client wrote, and writes things for it to check. */
public class Interop {
    static NetworkAccess network;
    static Crypto crypto;
    static Path dir;

    public static void main(String[] args) throws Exception {
        peergos.server.util.JavaInflate.init();
        crypto = Builder.initCrypto();
        network = Builder.buildJavaNetworkAccess(new URL("http://localhost:7777"), false, Optional.empty(), Optional.empty()).join();
        dir = Paths.get(args[1]);
        switch (args[0]) {
            case "check-rust" -> checkRust();
            case "java-setup" -> javaSetup();
            case "befriend" -> befriend();
            case "java-share" -> javaShare();
            case "check-rust-wrote" -> checkRustWrote();
            case "mfa-login" -> mfaLogin();
            default -> throw new IllegalStateException("unknown step " + args[0]);
        }
        System.exit(0);
    }

    static void check(boolean ok, String what) {
        if (! ok)
            throw new IllegalStateException("FAIL " + what);
        System.out.println("  ok   " + what);
    }

    static String hex(byte[] b) {
        StringBuilder s = new StringBuilder();
        for (byte x : b)
            s.append(String.format("%02x", x & 0xff));
        return s.toString();
    }

    static String sha(byte[] b) throws Exception {
        return hex(MessageDigest.getInstance("SHA-256").digest(b));
    }

    static byte[] readAll(AsyncReader r, long size) {
        byte[] res = new byte[(int) size];
        int done = 0;
        while (done < size) {
            int n = r.readIntoArray(res, done, (int) size - done).join();
            if (n <= 0)
                throw new IllegalStateException("short read");
            done += n;
        }
        return res;
    }

    static byte[] read(FileWrapper f) {
        return readAll(f.getInputStream(network, crypto, x -> {}).join(), f.getSize());
    }

    static FileWrapper get(UserContext ctx, String path) {
        return ctx.getByPath(path).join().orElseThrow(() -> new IllegalStateException("missing " + path));
    }

    static UserContext signIn(String user, String pw) {
        return UserContext.signIn(user, pw, req -> Futures.errored(new IllegalStateException("unexpected mfa")), network, crypto).join();
    }

    static UserContext signUpOrIn(String user, String pw) {
        try {
            return UserContext.signUp(user, pw, "", network, crypto).join();
        } catch (Exception e) {
            return signIn(user, pw);
        }
    }

    static FileWrapper upload(FileWrapper dir, String name, byte[] data) {
        return dir.uploadOrReplaceFile(name, AsyncReader.build(data), 0, data.length, network, crypto, x -> {}).join();
    }

    static FileWrapper mkdirIfAbsent(UserContext ctx, String parentPath, String name) {
        FileWrapper parent = get(ctx, parentPath);
        Optional<FileWrapper> existing = parent.getChild(name, crypto.hasher, network).join();
        if (existing.isPresent())
            return existing.get();
        parent.mkdir(name, network, false, Optional.empty(), crypto).join();
        return get(ctx, parentPath + "/" + name);
    }

    static byte[] pattern(int len, int seed) {
        byte[] b = new byte[len];
        for (int i = 0; i < len; i++)
            b[i] = (byte) (i * 31 + seed);
        return b;
    }

    static void checkRust() throws Exception {
        UserContext ri = signIn("ri", "ripass");
        for (String line : Files.readAllLines(dir.resolve("rust-expected.txt"))) {
            String[] parts = line.split(" ");
            check(sha(read(get(ri, parts[0]))).equals(parts[1]), parts[0] + " reads as the Rust client wrote it");
        }
        // the 4 MiB chunk file carries its chunk size and a BLAKE3 root
        FileWrapper b3 = get(ri, "/ri/interop/b3.bin");
        FileProperties props = b3.getFileProperties();
        check(props.chunkSize == Chunk.DEFAULT_SIZE, "b3.bin has the 4 MiB chunk size: " + props.chunkSize);
        String blake3 = Files.readString(dir.resolve("b3.blake3")).trim();
        check(hex(Blake3.hash(read(b3))).equals(blake3), "b3.bin's content hashes to the recorded BLAKE3");
        check(hex(props.treeHash.get().rootHash.hash).equals(blake3), "b3.bin's stored root is its BLAKE3 hash");

        // the zip the Rust client edited
        ZipReader zip = ZipReader.open(get(ri, "/ri/interop/archive.zip"), network, crypto).join();
        ZipEntry added = zip.getIndex().get("rust/added.txt").get();
        check(new String(readAll(zip.read(added).join(), added.size)).equals("added by the rust client"), "Rust-added zip entry reads in Java");
        ZipEntry readme = zip.getIndex().get("README.txt").get();
        check(readAll(zip.read(readme).join(), readme.size).length == 500, "Rust-renamed zip entry reads in Java");
        check(zip.getIndex().get("readme.txt").isEmpty() && zip.getIndex().get("empty").isEmpty(), "Rust-removed zip entries are gone");

        // the owner's view of the Rust link
        List<SecretLinkSummary> links = ri.getAllSecretLinks().join();
        String link = Files.readString(dir.resolve("rust-link.txt")).trim();
        SecretLinkSummary summary = links.stream().filter(s -> link.contains(s.linkString(ri.signer.publicKeyHash)))
                .findFirst().orElseThrow(() -> new IllegalStateException("Rust link not listed: " + links.size() + " links"));
        check(Arrays.asList(summary.paths()).equals(List.of("/ri/interop/small.txt", "/ri/interop/bulk")), "Java lists the Rust link's members: " + Arrays.toString(summary.paths()));
        check(summary.isWritable(), "Java sees the Rust link as writable");
        List<LinkMember> members = ri.getSecretLinkMembers(summary.props).join();
        check(members.size() == 2 && ! members.get(0).writable && members.get(1).writable, "Java reads the Rust link's payload");

        // open it as anyone with the link
        UserContext anon = UserContext.fromSecretLinkV2(link, () -> Futures.of(""), network, crypto).join();
        check(new String(read(get(anon, "/ri/interop/small.txt"))).equals("written by the rust client"), "Java opens the Rust link's read-only item");
        FileWrapper bulk = get(anon, "/ri/interop/bulk");
        check(bulk.isWritable(), "the Rust link's folder is writable in Java");
        upload(bulk, "via-rust-link.txt", "java wrote through a rust link".getBytes());
        System.out.println("check-rust done");
    }

    static void javaSetup() throws Exception {
        UserContext ji = signUpOrIn("ji", "jipass");
        FileWrapper root = mkdirIfAbsent(ji, "/ji", "jinterop");
        List<String> expected = new ArrayList<>();
        byte[] small = "written by the java client".getBytes();
        upload(root, "j-small.txt", small);
        expected.add("/ji/jinterop/j-small.txt " + sha(small));
        byte[] big = pattern(7 * 1024 * 1024 + 11, 4);
        upload(get(ji, "/ji/jinterop"), "j-big.bin", big);
        expected.add("/ji/jinterop/j-big.bin " + sha(big));
        mkdirIfAbsent(ji, "/ji/jinterop", "jlinkdir");
        mkdirIfAbsent(ji, "/ji/jinterop", "jshared");

        // a zip built with the Java ZipWriter, starting from an empty archive
        byte[] emptyZip = new byte[22];
        emptyZip[0] = 0x50; emptyZip[1] = 0x4b; emptyZip[2] = 0x05; emptyZip[3] = 0x06;
        upload(get(ji, "/ji/jinterop"), "java.zip", emptyZip);
        byte[] zipped = "zipped by the java client".getBytes();
        ZipWriter.NewEntry entry = new ZipWriter.NewEntry("docs/j.txt", zipped.length, LocalDateTime.now(),
                () -> Futures.of(AsyncReader.build(zipped)));
        ZipWriter.append(get(ji, "/ji/jinterop/java.zip"), List.of(entry), network, crypto, x -> {}).join();

        LinkProperties props = ji.createSecretLink(List.of("/ji/jinterop/j-small.txt", "/ji/jinterop/jlinkdir"),
                List.of("/ji/jinterop/jlinkdir"), Optional.empty(), Optional.empty(), "", false).join();
        Files.writeString(dir.resolve("java-link.txt"), ji.getLinkString(props));
        Files.writeString(dir.resolve("java-expected.txt"), String.join("\n", expected));
        System.out.println("java-setup done");
    }

    static void befriend() {
        UserContext ji = signIn("ji", "jipass");
        UserContext ri = signIn("ri", "ripass");
        ji.sendFollowRequest("ri", SymmetricKey.random()).join();
        for (FollowRequestWithCipherText req : ri.processFollowRequests().join())
            ri.sendReplyFollowRequest(req, true, true).join();
        ji.processFollowRequests().join();
        System.out.println("befriend done");
    }

    static void javaShare() {
        UserContext ji = signIn("ji", "jipass");
        ji.shareWriteAccessWith(Paths.get("/ji/jinterop/jshared"), Set.of("ri")).join();
        // write into the folder the Rust user shared with us
        FileWrapper shared = get(ji, "/ri/interop/shared");
        check(shared.isWritable(), "Rust's write share is writable in Java");
        upload(shared, "from-java.txt", "java wrote into rust's shared folder".getBytes());
        System.out.println("java-share done");
    }

    static void checkRustWrote() {
        UserContext ji = signIn("ji", "jipass");
        check(new String(read(get(ji, "/ji/jinterop/jshared/from-rust.txt"))).equals("rust wrote into java's shared folder"), "Rust wrote into Java's write share");
        check(new String(read(get(ji, "/ji/jinterop/jlinkdir/via-java-link.txt"))).equals("rust wrote through a java link"), "Java reads what Rust wrote through its link");
        System.out.println("check-rust-wrote done");
    }

    static void mfaLogin() throws Exception {
        String code = Files.readAllLines(dir.resolve("backup-codes.txt")).get(0);
        UserContext rm = UserContext.signIn("rm", "rmpass", req -> {
            MultiFactorAuthMethod m = req.methods.stream().filter(x -> x.type == MultiFactorAuthMethod.Type.BACKUP_CODES)
                    .findFirst().orElseThrow(() -> new IllegalStateException("no backup codes offered"));
            return Futures.of(new MultiFactorAuthResponse(m.credentialId, Either.a(code)));
        }, network, crypto).join();
        check(rm.username.equals("rm"), "Java signs in with a backup code the Rust client generated");
        System.out.println("mfa-login done");
    }
}

File: crates/peergos-fs/examples/interop.rs
//! Rust half of an interop run against a live server and the Java client. Each
//! step writes what it made (paths, sha256s, link strings) into a shared directory
//! for the Java half to check, or checks what the Java half wrote there.
//!
//!   cargo run -p peergos-fs --example interop -- <step> <shared-dir> [server]
//!
//! The Java half and the step order are in `interop/` at the repo root.

use peergos_core::mutable::{HttpMutablePointers, MutablePointers};
use peergos_core::{ContentAddressedStorage, HttpPoster, HttpStorage, ReqwestPoster};
use peergos_fs::archive::{self, EntrySource, NewEntry, ZipReader};
use peergos_fs::{FileUpload, FolderUpload, MultiFactorAuthResponse, UserContext};
use std::path::{Path, PathBuf};
use std::sync::Arc;

type Res<T> = Result<T, Box<dyn std::error::Error>>;

struct Net {
    poster: Arc<dyn HttpPoster>,
    store: Arc<dyn ContentAddressedStorage>,
    mutable: Arc<dyn MutablePointers>,
}

fn net(base: &str) -> Res<Net> {
    Ok(Net {
        poster: Arc::new(ReqwestPoster::new(base, false)?),
        store: Arc::new(HttpStorage::new(Arc::new(ReqwestPoster::new(base, false)?), true)),
        mutable: Arc::new(HttpMutablePointers::new(Arc::new(ReqwestPoster::new(base, false)?))),
    })
}

async fn sign_in(n: &Net, user: &str, pw: &str) -> Res<UserContext> {
    Ok(UserContext::sign_in(user, pw, None, n.poster.clone(), n.store.clone(), n.mutable.clone()).await?)
}

async fn sign_up_or_in(n: &Net, user: &str, pw: &str) -> Res<UserContext> {
    match UserContext::sign_up(user, pw, None, n.poster.clone(), n.store.clone(), n.mutable.clone()).await {
        Ok(c) => Ok(c),
        Err(_) => sign_in(n, user, pw).await,
    }
}

fn hex(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

fn sha(b: &[u8]) -> String {
    hex(&peergos_crypto::hash::sha256(b))
}

fn pattern(len: usize, seed: u8) -> Vec<u8> {
    (0..len).map(|i| ((i * 31) as u8).wrapping_add(seed)).collect()
}

fn check(ok: bool, what: &str) -> Res<()> {
    if ok {
        println!("  ok   {what}");
        Ok(())
    } else {
        Err(format!("FAIL {what}").into())
    }
}

async fn read_path(ctx: &UserContext, path: &str) -> Res<Vec<u8>> {
    Ok(ctx.get_by_path(path).await?.ok_or_else(|| format!("missing {path}"))?.read().await?)
}

/// Check every `path sha256` line of a file written by the Java half.
async fn check_expected(ctx: &UserContext, file: &Path) -> Res<()> {
    for line in std::fs::read_to_string(file)?.lines().filter(|l| !l.is_empty()) {
        let (path, want) = line.split_once(' ').ok_or("bad expected line")?;
        let got = sha(&read_path(ctx, path).await?);
        check(got == want, &format!("{path} reads as the Java client wrote it"))?;
    }
    Ok(())
}

async fn rust_setup(n: &Net, dir: &Path) -> Res<()> {
    let ctx = sign_up_or_in(n, "ri", "ripass").await?;
    let home = ctx.get_home().await?;
    let root = match home.child("interop").await? {
        Some(d) => d,
        None => home.mkdir("interop").await?,
    };
    let mut expected = Vec::new();
    let mut put = |path: &str, data: &[u8]| expected.push(format!("/ri/interop/{path} {}", sha(data)));

    let small = b"written by the rust client".to_vec();
    root.upload("small.txt", &small).await?;
    put("small.txt", &small);

    let big = pattern(11 * 1024 * 1024, 3);
    root.get_latest().await?.upload("big.bin", &big).await?;
    put("big.bin", &big);

    // a file at the 4 MiB BLAKE3 chunk size, as a newer client writes them
    let b3 = pattern(9 * 1024 * 1024 + 123, 5);
    let b3c = b3.clone();
    let r = root.get_latest().await?;
    peergos_fs::upload_file_streaming_at_chunk_size(
        r.capability(), "b3.bin", b3.len() as u64, peergos_fs::DEFAULT_CHUNK_SIZE, r.signer().cloned(),
        ctx.mirror_bat_id().as_ref(), move || Ok(std::io::Cursor::new(b3c.clone())), ctx.store(), ctx.mutable().as_ref(),
    )
    .await?;
    put("b3.bin", &b3);
    std::fs::write(dir.join("b3.blake3"), hex(&peergos_crypto::hash::blake3(&b3)))?;

    // a buffered subtree upload: bulk commits, split, with large blocks written ahead
    let mut files = Vec::new();
    for i in 0..40 {
        let d = pattern(90 * 1024, i as u8);
        put(&format!("bulk/s{i}.bin"), &d);
        files.push(FileUpload::from_bytes(format!("s{i}.bin"), d));
    }
    let large = pattern(6 * 1024 * 1024, 9);
    put("bulk/large.bin", &large);
    files.push(FileUpload::from_bytes("large.bin", large));
    let bulk = match root.get_latest().await?.child("bulk").await? {
        Some(d) => d,
        None => root.get_latest().await?.mkdir("bulk").await?,
    };
    bulk.upload_subtree(vec![FolderUpload { rel_path: vec![], files }]).await?;

    // a zip written by another tool, then edited here
    let fixture = std::fs::read(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/fixtures/sample.zip"))?;
    root.get_latest().await?.upload("archive.zip", &fixture).await?;
    let zip_file = ctx.get_by_path("interop/archive.zip").await?.ok_or("no zip")?;
    let added = b"added by the rust client".to_vec();
    archive::append(&zip_file, vec![NewEntry::file("rust/added.txt", added.len() as u64, 0, EntrySource::Bytes(Arc::new(added)))?]).await?;
    archive::rename(&zip_file, "readme.txt", "README.txt").await?;
    archive::remove(&zip_file, &["empty".to_string()], true).await?;

    // a link over two items, one writable
    let props = ctx
        .create_secret_link_to(
            &["interop/small.txt".to_string(), "interop/bulk".to_string()],
            &["interop/bulk".to_string()],
            "",
            None,
            None,
        )
        .await?;
    std::fs::write(dir.join("rust-link.txt"), ctx.secret_link_string(&props)?)?;
    if root.get_latest().await?.child("shared").await?.is_none() {
        root.get_latest().await?.mkdir("shared").await?;
    }
    std::fs::write(dir.join("rust-expected.txt"), expected.join("\n"))?;
    println!("rust-setup done");
    Ok(())
}

async fn rust_share(n: &Net) -> Res<()> {
    let ctx = sign_in(n, "ri", "ripass").await?;
    ctx.share_write_access("interop/shared", "ji").await?;
    println!("rust-share done: /ri/interop/shared is writable by ji");
    Ok(())
}

async fn check_java(n: &Net, dir: &Path) -> Res<()> {
    let ctx = sign_in(n, "ri", "ripass").await?;
    let ji = sign_in(n, "ji", "jipass").await?;
    check_expected(&ji, &dir.join("java-expected.txt")).await?;

    // the Java user's zip, built with its ZipWriter
    let zip_file = ji.get_by_path("/ji/jinterop/java.zip").await?.ok_or("no java zip")?;
    let zip = ZipReader::open(&zip_file).await?;
    check(zip.read_path("docs/j.txt").await? == b"zipped by the java client", "Java-written zip entry reads back")?;

    // the Java user's multi-item link
    let link = std::fs::read_to_string(dir.join("java-link.txt"))?;
    let anon = UserContext::from_secret_link(link.trim(), None, n.poster.clone(), n.store.clone(), n.mutable.clone()).await?;
    let mounts = anon.link_mount_paths();
    check(mounts.len() == 2, &format!("Java link opens with both items: {mounts:?}"))?;
    let a = anon.get_by_path("/ji/jinterop/j-small.txt").await?.ok_or("link item 1")?;
    check(a.read().await? == b"written by the java client", "Java link's read-only item reads")?;
    let wdir = anon.get_by_path("/ji/jinterop/jlinkdir").await?.ok_or("link item 2")?;
    check(wdir.is_writable(), "Java link's writable item is writable")?;
    wdir.upload("via-java-link.txt", b"rust wrote through a java link").await?;
    check(
        read_path(&ji, "/ji/jinterop/jlinkdir/via-java-link.txt").await? == b"rust wrote through a java link",
        "Rust wrote through the Java link",
    )?;

    // what the Java user wrote into our write-shared folder
    check(read_path(&ctx, "interop/shared/from-java.txt").await? == b"java wrote into rust's shared folder", "Java wrote into Rust's write share")?;
    // write into the folder the Java user shared with us
    let jshared = ctx.get_by_path("/ji/jinterop/jshared").await?.ok_or("no jshared")?;
    check(jshared.is_writable(), "Java's write share is writable for Rust")?;
    jshared.upload("from-rust.txt", b"rust wrote into java's shared folder").await?;
    // what the Java client wrote through our link
    check(read_path(&ctx, "interop/bulk/via-rust-link.txt").await? == b"java wrote through a rust link", "Java wrote through the Rust link")?;
    println!("check-java done");
    Ok(())
}

async fn rust_mfa(n: &Net, dir: &Path) -> Res<()> {
    let ctx = sign_up_or_in(n, "rm", "rmpass").await?;
    let totp = ctx.enroll_totp().await?;
    let codes = ctx.generate_backup_codes().await?;
    std::fs::write(dir.join("backup-codes.txt"), codes.formatted().join("\n"))?;
    std::fs::write(dir.join("totp.txt"), totp.encode())?;
    // sign in again answering with the last backup code
    let last = codes.codes.last().unwrap().clone();
    let responder = move |req: &peergos_fs::MultiFactorAuthRequest| {
        let m = req.backup_codes_method().ok_or_else(|| peergos_core::Error::Protocol("no backup codes offered".into()))?;
        Ok(MultiFactorAuthResponse::new_backup_code(m.credential_id.clone(), &last))
    };
    let again = UserContext::sign_in("rm", "rmpass", Some(&responder), n.poster.clone(), n.store.clone(), n.mutable.clone()).await?;
    check(again.username() == Some("rm"), "Rust signs in with a backup code")?;
    let totp_responder = |req: &peergos_fs::MultiFactorAuthRequest| {
        let m = req.totp_method().ok_or_else(|| peergos_core::Error::Protocol("no totp offered".into()))?;
        Ok(MultiFactorAuthResponse::new_totp(m.credential_id.clone(), totp.current_code()))
    };
    UserContext::sign_in("rm", "rmpass", Some(&totp_responder), n.poster.clone(), n.store.clone(), n.mutable.clone()).await?;
    check(true, "Rust signs in with TOTP")?;
    println!("rust-mfa done");
    Ok(())
}

#[tokio::main]
async fn main() -> Res<()> {
    let args: Vec<String> = std::env::args().collect();
    let step = args.get(1).ok_or("step")?.clone();
    let dir = PathBuf::from(args.get(2).ok_or("shared dir")?);
    let base = args.get(3).cloned().unwrap_or_else(|| "http://localhost:7777/".into());
    std::fs::create_dir_all(&dir)?;
    let n = net(&base)?;
    match step.as_str() {
        "rust-setup" => rust_setup(&n, &dir).await,
        "rust-share" => rust_share(&n).await,
        "check-java" => check_java(&n, &dir).await,
        "rust-mfa" => rust_mfa(&n, &dir).await,
        other => Err(format!("unknown step {other}").into()),
    }
}
